### PR TITLE
Disallow outer joins `ON TRUE` with ref & dist tables when ref table is outer relation

### PR DESCRIPTION
**Pull request by onurctirtir**
Fixes https://github.com/citusdata/citus/issues/4094.

DESCRIPTION: Disallows `ON TRUE` outer joins with ref & dist tables when ref table is outer rel

------

In `ContextCoversJoinRestriction`, we were calling `LeftListIsSubset` in a
wrong way so were throwing off some join restrictions and we sometimes
were not able detect such unsupported joins with distributed tables and
reference tables where reference table is the outer relation in the join.
So, we were prunning join to 0 shards.
(See https://github.com/citusdata/citus/issues/4094#issuecomment-708991177)

This pr actually fixes that with e597145a79d6ec4c527684777364ebede3274cd0.

But after e597145a79d6ec4c527684777364ebede3274cd0, we had a regression test output change in existing
regression tests.

Problematic query was in `multi_subquery_complex_reference_clause.sql`:
```sql
SELECT count(*) FROM users_ref_test_table ref1
LEFT JOIN users_ref_test_table ref2 on ref1.id = ref2.id
LEFT JOIN user_buy_test_table ON true;
```
And to fix, if query has not only inner joins, then we do not throw off
`joinRestriction` if `joinRestrictInfoList` that `joinRestriction` stores is
empty at 42c9136ab09a394870cd6709d0e2edab4363d071.
The methodology is, `LeftListIsSubset` returns true even if 
`joinRestriction->joinRestrictInfoList` is empty. However, `joinRestriction` 
might itself still contain useful information for outer joins.

----

(For easier review, split the work into commits, I will remove reference to issue
number in commit message and squash commits before merging)

----

TODO:
- [x] Add regression tests from #4094 
- [x] Move `hasSemi` field down to `JoinRestrictionContext` in another pr and move `hasOnlyInnerJoin` field in this pr (https://github.com/citusdata/citus/pull/4256)

----

Also, to understand how this pr might affect performance of such `ON TRUE`  outer joins, see https://github.com/citusdata/citus/pull/4255#issuecomment-712153217

### CODE REVIEW

File: src/backend/distributed/planner/distributed_planner.c
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:1746]: I think we should move both hasOnlyInnerJoin and hasSemiJoin  to `JoinRestrictionContext`. With  that, we don't have to export/call `CurrentPlannerRestrictionContext`. (And, it has many implications anyway, we cannot do it).

File: src/backend/distributed/planner/distributed_planner.c
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:115]: we should add all the tests mentioned in https://github.com/citusdata/citus/issues/4094 (if not more)

File: src/backend/distributed/planner/distributed_planner.c
Comment (onurctirtir) [reply src/backend/distributed/planner/distributed_planner.c]: yeah, I added that to TODO's, will open a seperate pr for that

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:2059]: This part seems complicated. Can we split the two checks?
```
		bool joinIsOnTrue = list_length(joinRestrictInfoListInTest) == 0;
		bool hasOnlyInnerJoin = joinRestrictionContext->hasOnlyInnerJoin;
		if (joinIsOnTrue && !hasOnlyInnerJoin)
		{
			/*
			 * If the join doesn't have a restriction (e.g., ON (true)) and
			 * the planner is aware of at least one non-inner JOIN (e.g., outer/semi
			 * joins), we should not eliminiate joinRestrictInfoListInTest. It can
			 * be useful for detecting not supported outer-join checks even if it doesn't
			 * help for colocation checks.
			 */
			continue;
		}
```

**Review by onderkalaci — COMMENTED** [review]
> yeah, I added that to TODO's, will open a seperate pr for that

Alright, we cannot merge this PR before that lands because we cannot liberally call `CurrentPlannerRestrictionContext ()`, we should rely on the `plannerRestrictionInfo` that is passed to the function

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onurctirtir) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:1900]: maybe a better comment ?

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onurctirtir) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:1903]: those can go into a function

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onurctirtir) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:2061]: I wonder if this is specific for "on true" or should we give it a bit more generic name

**Review by onurctirtir — COMMENTED** [review]
@onderkalaci, I will add regression tests as noted in TODO, I will also reconsider below

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (marcocitus) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2061]: perhaps just something like `hasJoinRestriction = list_length(joinRestrictInfoListInTest) > 0`

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:2070]: why we remove this comment? Can't we adjust?   I think having at least the following would be nice: ` equal() function is implemented for the struct.`

Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2070]: Both we already mention `equal` function in `LeftListIsSubset` and that we use `LeftListIsSubset` instead of `list_difference`, I thought it is already clear what we do here.  What do you think ?

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:1909]: or just removing it as it doesn't seem to add any value

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (marcocitus) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:2068]: this should be `> 0` and the if statement should use `!hasJoinRestriction`

Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2068]: oh this is true, I will apply that

File: src/backend/distributed/planner/distributed_planner.c
Comment (onurctirtir) [reply src/backend/distributed/planner/distributed_planner.c]: added in 6e99005, now thinking of a bit more complicated regression tests

Comment (onderkalaci) [reply src/backend/distributed/planner/distributed_planner.c]: we should consider adding all the examples listed in https://github.com/citusdata/citus/issues/4094, especially the sqlancer one discussed on the description

Comment (onurctirtir) [reply src/backend/distributed/planner/distributed_planner.c]: thanks, I will add those tests too

Comment (onurctirtir) [reply src/backend/distributed/planner/distributed_planner.c]: added in https://github.com/citusdata/citus/pull/4255/commits/abe5062d32fc3a539dba3084b55ca0c713aaa644

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:2099]: I think this function is not very much necessary. The name is hard to follow etc. Can't we inline it back?

Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2099]: I'm okay with removing it, will do

File: src/test/regress/sql/sqlancer_failures.sql
Comment (onderkalaci) [inline src/test/regress/sql/sqlancer_failures.sql:125]: Let's have more tests:
```SQL





-- unsupported outer JOIN inside a subquery in WHERE clause
SELECT * FROM distributed_table WHERE buy_count > (
SELECT count(*) FROM distributed_table a
LEFT JOIN reference_table b ON (true)
RIGHT JOIN reference_table c ON (false));


-- unsupported outer JOIN via subqueries
SELECT count(*) FROM (SELECT *, random() FROM distributed_table) as a
LEFT JOIN (SELECT *, random() FROM reference_table) as b ON (true)
RIGHT JOIN (SELECT *, random() FROM reference_table) as c ON (false);


-- unsupported outer JOIN in a sublevel subquery
SELECT 
   count(*) 
FROM  
   (
      SELECT  a.* FROM distributed_table a JOIN distributed_table b USING (user_id) 
   ) as bar
JOIN
   (
      SELECT a.* FROM distributed_table a
      LEFT JOIN reference_table b ON (true)
      RIGHT JOIN reference_table c ON (true)
   ) as unsupported_join
ON (true);

-- unsupported outer JOIN in a sublevel INNER JOIN
SELECT
 unsupported_join.* 
FROM 
   (distributed_table a
   LEFT JOIN reference_table b ON (true)
   RIGHT JOIN reference_table c ON (true)) as unsupported_join (x,y,z,t,e,f,q)
JOIN
   (reference_table d JOIN reference_table e ON(true)) ON (true);

-- unsupported outer JOIN in a sublevel LEFT JOIN
SELECT
 unsupported_join.* 
FROM 
   (distributed_table a
   LEFT JOIN reference_table b ON (true)
   RIGHT JOIN reference_table c ON (true)) as unsupported_join 
LEFT JOIN
   (reference_table d JOIN reference_table e ON(true)) ON (true);


-- unsupported outer JOIN in a sublevel RIGHT JOIN
SELECT
 unsupported_join.* 
FROM 
   (distributed_table a
   LEFT JOIN reference_table b ON (true)
   RIGHT JOIN reference_table c ON (false)) as unsupported_join 
RIGHT JOIN
   (reference_table d JOIN reference_table e ON(true)) ON (true);





```

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2076]: I wonder if we should introduce a function named as `list_empty` to listutils in another pr to make such checks shorter and safer

File: src/test/regress/sql/sqlancer_failures.sql
Comment (onurctirtir) [reply src/test/regress/sql/sqlancer_failures.sql:125]: thanks for additional test cases

**Comment by onurctirtir** [comment]
For future reference, looking into below query:

```sql
select count(*) from distributed_table as dist LEFT
JOIN reference_table u1 ON (true) LEFT
JOIN reference_table u2 ON (true) LEFT
JOIN reference_table u3 ON (true) LEFT
.
.
.
JOIN users_ref_test_table u50 ON (true);
```

While `RemoveDuplicateJoinRestrictions` reduces number of join restrictions from 100 to 2 on master (6e493624af4eaada536eb1201dd0d82ef53f4b9c), we are not able to eliminate any join restrictions in this pr as all of them are `ON TRUE` joins.

Time it takes to execute `EXPLAIN` for that query doesn't change between master and this pr.

**Review by onderkalaci — APPROVED** [review]
Thanks, this looks good. I've done various tests and the new set of queries added makes sense. Please do not forget to mention the before/after planning time changes for joins that are affected by this PR

**Pull request merged**
